surface/include/pcl/surface/impl/mls.hpp

```diff
@@ -673,45 +534,363 @@ pcl::MovingLeastSquares<PointInT, PointOutT>::performUpsampling (PointCloudOut &
+
+  return result;
+}
+
+pcl::MLSProjectionResults pcl::projectPointToMLSPlane(const double u, const double v, const pcl::MLSResult &mls_result)
+{
+  MLSProjectionResults result;
+  result.u = u;
+  result.v = v;
+  result.normal = mls_result.plane_normal;
+  result.point = mls_result.mean + u * mls_result.u_axis + v * mls_result.v_axis;
+
+  return result;
+}
+
+pcl::MLSProjectionResults pcl::projectPointSimpleToPolynomialSurface (const double u, const double v, const pcl::MLSResult &mls_result)
+{
+  MLSProjectionResults result;
+  double w = 0;
+
+  result.normal = mls_result.plane_normal;
+
+  if (mls_result.polynomial_fit && mls_result.c_vec.size () >= (mls_result.order + 1) * (mls_result.order + 2) / 2 && pcl_isfinite (mls_result.c_vec[0]))
+  {
+    PolynomialPartialDerivative d = getPolynomialPartialDerivative (u, v, mls_result);
+    w = d.z;
+    result.normal -= (d.z_u * mls_result.u_axis + d.z_v * mls_result.v_axis);
+    result.normal.normalize();
+  }
+
+  result.point = mls_result.mean + u * mls_result.u_axis + v * mls_result.v_axis + w * mls_result.plane_normal;
+
+  return result;
+}
+
+template <typename PointT>
+void pcl::computeMLSSurface (const pcl::PointCloud<PointT> &cloud,
+                             int index,
+                             const std::vector<int> &nn_indices,
+                             double search_radius,
+                             pcl::MLSResult &mls_result,
+                             bool polynomial_fit,
+                             int polynomial_order,
+                             boost::function<double(const double)> weight_func)
+{
+  // Compute the plane coefficients
+  EIGEN_ALIGN16 Eigen::Matrix3d covariance_matrix;
+  Eigen::Vector4d xyz_centroid;
+
+  // Estimate the XYZ centroid
+  pcl::compute3DCentroid (cloud, nn_indices, xyz_centroid);
+
+  // Compute the 3x3 covariance matrix
+  pcl::computeCovarianceMatrix (cloud, nn_indices, xyz_centroid, covariance_matrix);
+  EIGEN_ALIGN16 Eigen::Vector3d::Scalar eigen_value;
+  EIGEN_ALIGN16 Eigen::Vector3d eigen_vector;
+  Eigen::Vector4d model_coefficients;
+  pcl::eigen33 (covariance_matrix, eigen_value, eigen_vector);
+  model_coefficients.head<3> ().matrix () = eigen_vector;
+  model_coefficients[3] = 0;
+  model_coefficients[3] = -1 * model_coefficients.dot (xyz_centroid);
+
+  // Projected query point
+  mls_result.valid = true;
+  mls_result.query_point = cloud.points[index].getVector3fMap ().template cast<double> ();
+  double distance = mls_result.query_point.dot (model_coefficients.head<3> ()) + model_coefficients[3];
+  mls_result.mean = mls_result.query_point - distance * model_coefficients.head<3> ();
+
+  mls_result.curvature = static_cast<float> (covariance_matrix.trace ());
+  // Compute the curvature surface change
+  if (mls_result.curvature != 0)
+    mls_result.curvature = fabsf (float (eigen_value / double (mls_result.curvature)));
```

Comment:
Why is the curvature casted to double?

Reply:
Also, please always use `std::abs` et al., we had problems with MSVC if I remember correctly.

Reply:
I believe that is how it was in the original implementation.

Reply:
Nevertheless. PCL is written C++ and we should always prefer C++ library functions over their C-style counterparts. Besides, as I wrote, we occasionally get bug reports about MSVC choking at some of these functions (#848, #1548).

Reply:
Sorry, my comment was in reference to the cast to double.